mender-image-prepare.sh

```diff
@@ -0,0 +1,535 @@
+}
+
+detach_mappings() {
+  for mapping in ${mappings[@]}
+  do
+    map_dev=/dev/mapper/"$mapping"
+    is_mounted=`grep ${map_dev} /proc/self/mounts | wc -l`
+    if [ ${is_mounted} -ne 0 ]; then
+      umount $map_dev
+    fi
+  done
+
+  mapper=${mappings[0]%p*}
+  kpartx -d /dev/$mapper
+  losetup -d /dev/$mapper
+}
+
+map_image() {
+  loopline=$(kpartx -v -a $image | grep 'loop' | cut -d' ' -f3)
+  local loopmap=${loopline}
+  local loopdev=${loopline%p*}
+  partprobe /dev/$loopdev
+
+  mkdir -p $BASE_DIR/mnt/rootfs
+  mount /dev/mapper/"${loopmap}" $BASE_DIR/mnt/rootfs
+  echo ${loopline%p*}
+}
+
+mount_target_dev() {
+  partitions=($(fdisk -l ${target_dev} | grep '^/dev' | cut -d' ' -f1))
+
+  number_of_parts=${#partitions[@]}
+
+  [[ $number_of_parts -eq 4 ]] || { echo "Error: incorect device partitioning."; exit 1; }
+
+  echo "Mount target device's $number_of_parts partitions..."
+
+  for partition in "${partitions[@]}"
+  do
+    part_no=$(get_part_number_from_device $partition)
+    path=$BASE_DIR/mnt/${part_names[${part_no} - 1]}
+    mkdir -p ${path}
+
+    mount $partition $path
+  done 
+}
+
+mount_target_dev_sdimg() {
+  partitions=($(fdisk -l ${target_dev} | grep '^mender' | cut -d' ' -f1))
+
+  number_of_parts=${#partitions[@]}
+
+  [[ $number_of_parts -eq 4 ]] || { echo "Error: incorect device partitioning."; exit 1; } 
+
+  echo "Mount target device's $number_of_parts partitions..."
+
+  dev=$(losetup --show -f -P $target_dev)
+
+  loopsdimg=$dev
+
+  for partition in "${dev}"?*; do
+    echo "Mounting target partition: $partition"
+
+    part_no=$(get_part_number_from_device $partition)
+    path=$BASE_DIR/mnt/${part_names[${part_no} - 1]}
+
+    mkdir -p ${path}
+    mount ${partition} ${path}
+  done
+}
+
+set_uenv() {
+  [ ! -d "./mnt/boot" ] && { echo "Error: check mounted boot location."; exit 1; }
+  cat <<- 'EOF' > mnt/boot/uEnv.txt
+	loadaddr=0x82000000
+	fdtaddr=0x88000000
+	rdaddr=0x88080000
+
+	initrd_high=0xffffffff
+	fdt_high=0xffffffff
+
+	loadximage=echo debug: [/boot/vmlinuz-${uname_r}] ... ; load mmc 0:2 ${loadaddr} /boot/vmlinuz-${uname_r}
+	loadxfdt=echo debug: [/boot/dtbs/${uname_r}/${fdtfile}] ... ;load mmc 0:2 ${fdtaddr} /boot/dtbs/${uname_r}/${fdtfile}
+	loadxrd=echo debug: [/boot/initrd.img-${uname_r}] ... ; load mmc 0:2 ${rdaddr} /boot/initrd.img-${uname_r}; setenv rdsize ${filesize}
+	loaduEnvtxt=load mmc 0:2 ${loadaddr} /boot/uEnv.txt ; env import -t ${loadaddr} ${filesize};
+	check_dtb=if test -n ${dtb}; then setenv fdtfile ${dtb};fi;
+	loadall=run loaduEnvtxt; run check_dtb; run loadximage; run loadxrd; run loadxfdt;
+
+	mmcargs=setenv bootargs console=tty0 console=${console} ${optargs} ${cape_disable} ${cape_enable} root=/dev/mmcblk0p2 rootfstype=${mmcrootfstype} ${cmdline}
+
+	uenvcmd=run loadall; run mmcargs; echo debug: [${bootargs}] ... ; echo debug: [bootz ${loadaddr} ${rdaddr}:${rdsize} ${fdtaddr}] ... ; bootz ${loadaddr} ${rdaddr}:${rdsize} ${fdtaddr};
```

Comment:
What is the plan with `uenv.txt`? Ideally we should let U-Boot run `distro_bootcmd` which should run GRUB through `bootefi`, which in turn means we should not need any custom U-Boot commands at all. But one step at a time, I guess, just curious.

Reply:
In final version I think there will be no such customized uEnv.txt file. Since there is no integration with GRUB right now I just wanted to have something working, so that the kernel and filesystem starts.

Reply:
Ok, makes sense.